packages/workspace/src/browser/workspace-frontend-contribution.ts

```diff
@@ -262,28 +262,60 @@ export class WorkspaceFrontendContribution implements CommandContribution, Keybi
      * This is the generic `Open` method. Opens files and directories too. Resolves to the opened URI.
      * Except when you are on either Windows or Linux `AND` running in electron. If so, it opens a file.
      */
-    protected async doOpen(): Promise<URI | undefined> {
+    protected async doOpen(): Promise<MaybeArray<URI> | undefined> {
         if (!isOSX && this.isElectron()) {
             return this.doOpenFile();
         }
         const [rootStat] = await this.workspaceService.roots;
         const destinationUri = await this.fileDialogService.showOpenDialog({
             title: WorkspaceCommands.OPEN.dialogLabel,
             canSelectFolders: true,
-            canSelectFiles: true
+            canSelectFiles: true,
+            canSelectMany: true
         }, rootStat);
-        if (destinationUri && this.getCurrentWorkspaceUri()?.toString() !== destinationUri.toString()) {
-            const destination = await this.fileService.resolve(destinationUri);
-            if (destination.isDirectory) {
-                this.workspaceService.open(destinationUri);
+        if (destinationUri) {
+            if (Array.isArray(destinationUri)) {
```

Comment:
On a general note, I think APIs that return `T | T[]` don't make sense in many cases: if you care about the "single element" case, you can still check for `length === 1` and if you don't care, you don't have to handle the case.

Reply:
Thanks for the hint to simplify algorithm! Indeed, that's better to handle all cases at once. 
I made explicit in the signature of the doOpen... methods that an array of elements will be returned, getting rid of the MaybeArray. I kept however the signature of the showOpenDialog service with this MaybeArray, to avoid compatibility issues.